Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp

```diff
@@ -902,22 +902,21 @@ void NetworkConnectionToWebProcess::domCookiesForHost(const URL& url, Completion
 
 #if HAVE(COOKIE_CHANGE_LISTENER_API)
 
-void NetworkConnectionToWebProcess::subscribeToCookieChangeNotifications(const String& host)
+void NetworkConnectionToWebProcess::subscribeToCookieChangeNotifications(const URL& url, const URL& firstParty, WebCore::FrameIdentifier frameID, WebCore::PageIdentifier pageID, WebCore::ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, CompletionHandler<void(bool)>&& completionHandler)
 {
-    ASSERT(!m_hostsWithCookieListeners.contains(host));
-    m_hostsWithCookieListeners.add(host);
+    MESSAGE_CHECK_COMPLETION(protectedNetworkProcess()->allowsFirstPartyForCookies(m_webProcessIdentifier, firstParty), completionHandler(false));
 
-    if (auto* networkStorageSession = storageSession())
-        networkStorageSession->startListeningForCookieChangeNotifications(*this, host);
+    bool startedListening = false;
+    if (CheckedPtr networkStorageSession = storageSession())
+        startedListening = networkStorageSession->startListeningForCookieChangeNotifications(*this, url, firstParty, frameID, pageID, shouldRelaxThirdPartyCookieBlocking);
+
+    completionHandler(startedListening);
 }
 
 void NetworkConnectionToWebProcess::unsubscribeFromCookieChangeNotifications(const String& host)
 {
-    bool removed = m_hostsWithCookieListeners.remove(host);
-    ASSERT_UNUSED(removed, removed);
-
     if (auto* networkStorageSession = storageSession())
-        networkStorageSession->stopListeningForCookieChangeNotifications(*this, HashSet<String> { host });
+        networkStorageSession->stopListeningForCookieChangeNotifications(*this);
```

Comment:
Why is this change OK? `stopListeningForCookieChangeNotifications()` will unregistered ALL listeners now, not just the one for the given host. The host is now ignored in the new code which doesn't seem to make sense.

Reply:
This change is not ok. I asked for removing `m_hostsWithCookieListeners` because `NetworkStorageSession` already keeps a map of { host, `NetworkConnectionToWebProcess`s}, and it's unnecessary to keep information about in two places (`NetworkStorageSession` and `NetworkConnectionToWebProcess`).
i.e. `networkStorageSession->stopListeningForCookieChangeNotifications(*this, m_hostsWithCookieListeners);` can be done without `m_hostsWithCookieListeners` because it just means to remove all entries for the `NetworkConnectionToWebProcess` to `NetworkStorageSession`.

But I didn't ask for removing `host` from `stopListeningForCookieChangeNotifications`. `networkStorageSession->stopListeningForCookieChangeNotifications(*this, HashSet<String> { host });` is about to remove one entry to `NetworkConnectionToWebProcess `, and `NetworkStorageSession ` need the host information. 

I am ok with not removing `m_hostsWithCookieListeners` in this patch if it means too much work. We could do it in followup with test to verify we don't remove subscription accidentally. (Note if we keep `m_hostsWithCookieListeners`, we only want to update it when `startedListening` is true).